### PR TITLE
New Fixer - CurrencyRates ValueObject

### PR DESCRIPTION
**Pull request by vikkio88**
- moved call to rate to valueObject class
- moved fixer provider to `FreeFixerProvider`
- extended `FreeFixerProvider` to a `FixerProvider` that will use the paid for api.
- added formatter for CurrencyLayerRates

### CODE REVIEW

File: spec/Actions/RatesGetSpec.php [hunk 1]
```diff
@@ -21,7 +22,8 @@ function let(
         Request $request,
         Response $response,
         Route $route
-    ) {
+    )
```
Comment (mheki) [inline spec/Actions/RatesGetSpec.php]: can we revert this change? 
I know we don't check specs against psr-2 but for the sake of consistency I'd leave brackets in multiline methods as they were

File: spec/CurrencyRates/Provider/CurrencyLayer/CurrencyLayerProviderSpec.php [hunk 2]
```diff
@@ -30,7 +29,7 @@ function it_is_a_currency_provider()
         $this->shouldBeAnInstanceOf(CurrencyRatesProvider::class);
     }
 
-    function it_returns_an_array_if_getRates_is_called(HttpClient $httpClient, Config $config)
+    function it_returns_an_currency_rates_if_getRates_is_called(HttpClient $httpClient, Config $config)
```
Comment (mheki) [inline spec/CurrencyRates/Provider/CurrencyLayer/CurrencyLayerProviderSpec.php]: `it_returns_currency_rates` is descriptive enough

File: src/Actions/RatesGet.php [hunk 3]
```diff
@@ -33,7 +33,7 @@ protected function action(): array
             $currency = new Currency($this->get('baseCurrency', ''));
             $this->guard->assertAllowedCurrency($currency);
 
-            return $this->ratesProvider->getRates($currency);
+            return $this->ratesProvider->getRates($currency)->getPayload();
```
Comment (mheki) [inline src/Actions/RatesGet.php]: can we introduce local variable for the readability, please?
```php
$currencyRates = $this->ratesProvider->getRates($currency);

return $currencyRates->getPayload();
```

File: src/CurrencyRates/CurrencyRates.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace CurrencyPlatform\CurrencyRates;
+
+class CurrencyRates
+{
+    /** @var array */
+    protected $rates;
```
Comment (mheki) [inline src/CurrencyRates/CurrencyRates.php]: Can we use final classes with private fields for VOs, please?

File: src/CurrencyRates/Provider/Fixer/FixerRatesProvider.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+use Noodlehaus\Config;
+
+class FixerRatesProvider extends FreeFixerRatesProvider
+{
+    /**
+     * @var Config
+     */
+    protected $config;
+
+    protected $ratesUrl = 'http://data.fixer.io/api/latest';
```
Comment (mheki) [inline src/CurrencyRates/Provider/Fixer/FixerRatesProvider.php]: Can we get url from config, please?

File: src/CurrencyRates/Provider/Fixer/FreeFixerRatesProvider.php [hunk 6]
```diff
@@ -1,45 +1,41 @@
...
-     * @var Config
-     */
-    private $config;
-    private $cachedTimeInSeconds = 3600;
-    private $ratesUrl = 'https://api.fixer.io/latest';
-    private $storagePath = 'storage';
+    protected $httpClient;
+    protected $cachedTimeInSeconds = 3600;
+    protected $ratesUrl = 'https://api.fixer.io/latest';
+    protected $storagePath = 'storage';
```
Comment (mheki) [inline src/CurrencyRates/Provider/Fixer/FreeFixerRatesProvider.php]: Would be good to get these config values from config as well

File: src/CurrencyRates/Provider/CurrencyLayer/Rates/CurrencyLayerRates.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * This function is used to clean up the output of CurrencyLayer rates
+     * which quotes are shown as follows
+     * "USDEUR":0.810603,
+     * "USDGBP":0.719911,
+     * "USDCAD":1.285809,
+     * "USDPLN":3.416196
+     * @return array
+     */
+    public function getRates(): array
```
Comment (mheki) [inline src/CurrencyRates/Provider/CurrencyLayer/Rates/CurrencyLayerRates.php]: I don't think we need this class. 
It's better to keep `CurrencyRates` final and move this logic into provider.
This way adding/removing provider means you only add/remove one class.

File: spec/Actions/RatesGetSpec.php (hunk 1 above)
Comment (vikkio88) [reply spec/Actions/RatesGetSpec.php]: yeah, I have the habit of clicking Ctrl+Alt+L and does all of this. will do

File: src/CurrencyRates/CurrencyRates.php (hunk 4 above)
Comment (vikkio88) [reply src/CurrencyRates/CurrencyRates.php]: I am extending it (CurrencyLayerRates does some elaboration)

File: src/CurrencyRates/Provider/Fixer/FreeFixerRatesProvider.php (hunk 6 above)
Comment (vikkio88) [reply src/CurrencyRates/Provider/Fixer/FreeFixerRatesProvider.php]: as the url those are specific for this implementation, I dont think give us much benefit to move them in a config

File: src/CurrencyRates/Provider/CurrencyLayer/Rates/CurrencyLayerRates.php (hunk 7 above)
Comment (vikkio88) [reply src/CurrencyRates/Provider/CurrencyLayer/Rates/CurrencyLayerRates.php]: I thing is better than this class is given the logic to treat the rates coming from the Provider

File: src/CurrencyRates/Provider/Fixer/FixerRatesProvider.php (hunk 5 above)
Comment (vikkio88) [reply src/CurrencyRates/Provider/Fixer/FixerRatesProvider.php]: as per https://github.com/rushpartners/currency-platform/pull/5#discussion_r175029930

File: spec/Actions/RatesGetSpec.php (hunk 1 above)
Comment (vikkio88) [reply spec/Actions/RatesGetSpec.php]: Anyhow, we do check for psr2 and that is compliant

File: src/CurrencyRates/Provider/Fixer/FreeFixerRatesProvider.php (hunk 6 above)
Comment (mheki) [reply src/CurrencyRates/Provider/Fixer/FreeFixerRatesProvider.php]: OK, let's leave it as is and if there's a problem we'll come back to it later
